File: src/main/scala/com/github/shadowsocks/utils/Console.scala
/*
 * Shadowsocks - A shadowsocks client for Android
 * Copyright (C) 2014 <max.c.lv@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 *                            ___====-_  _-====___
 *                      _--^^^#####//      \\#####^^^--_
 *                   _-^##########// (    ) \\##########^-_
 *                  -############//  |\^^/|  \\############-
 *                _/############//   (@::@)   \\############\_
 *               /#############((     \\//     ))#############\
 *              -###############\\    (oo)    //###############-
 *             -#################\\  / VV \  //#################-
 *            -###################\\/      \//###################-
 *           _#/|##########/\######(   /\   )######/\##########|\#_
 *           |/ |#/\#/\#/\/  \#/\##\  |  |  /##/\#/  \/\#/\#/\#| \|
 *           `  |/  V  V  `   V  \#\| |  | |/#/  V   '  V  V  \|  '
 *              `   `  `      `   / | |  | | \   '      '  '   '
 *                               (  | |  | |  )
 *                              __\ | |  | | /__
 *                             (vvv(VVV)(VVV)vvv)
 *
 *                              HERE BE DRAGONS
 *
 */

package com.github.shadowsocks.utils

import java.util

import eu.chainfire.libsuperuser.Shell
import eu.chainfire.libsuperuser.Shell.{Builder, SU}

object Console {

  private def openShell(): Shell.Interactive = {
    val builder = new Builder()
    builder
      .useSH()
      .setWatchdogTimeout(10)
      .open()
  }

  private def openRootShell(context: String): Shell.Interactive = {
    val builder = new Builder()
    builder
      .setShell(SU.shell(0, context))
      .setWantSTDERR(true)
      .setWatchdogTimeout(10)
      .open()
  }

  def runCommand(commands: Array[String]) {
    val shell = openShell()
    shell.addCommand(commands, 0, ((commandCode: Int, exitCode: Int, output: util.List[String]) =>
      if (exitCode < 0) shell.close()): Shell.OnCommandResultListener)
    shell.waitForIdle()
    shell.close()
  }

  def runRootCommand(commands: String*): String = runRootCommand(commands.toArray)
  def runRootCommand(commands: Array[String]): String = {
    val shell = openRootShell("u:r:init_shell:s0")
    val sb = new StringBuilder
    shell.addCommand(commands, 0, ((_, exitCode, output) => {
      if (exitCode < 0) {
        shell.close()
      } else {
        import scala.collection.JavaConversions._
        output.foreach(line => sb.append(line).append('\n'))
      }
    }): Shell.OnCommandResultListener)
    if (shell.waitForIdle()) {
      shell.close()
      sb.toString()
    }
    else {
      shell.close()
      null
    }
  }

  def isRoot: Boolean = SU.available()
}

File: app/src/main/java/yyf/shadowsocks/utils/Console.java
package yyf.shadowsocks.utils;

/**
 * Created by yyf on 2015/6/26.
 */
import java.util.List;

import eu.chainfire.libsuperuser.Shell;
import eu.chainfire.libsuperuser.Shell.*;

//import java.util

public class Console {


    static private Shell.Interactive openShell() {
        Builder builder = new Builder();
        return builder.useSH().setWatchdogTimeout(10).open();
    }

    static private Shell.Interactive openRootShell(String context) {
        Builder builder = new Builder();
        return builder.setShell(SU.shell(0, context)).setWantSTDERR(true).setWatchdogTimeout(10).open();
    }

    static public void runCommand(String command) {
        final Interactive shell = openShell();
        shell.addCommand(command, 0, new OnCommandResultListener() {
            @Override
            public void onCommandResult(int commandCode, int exitCode, List<String> output) {
                if (exitCode < 0) {
                    shell.close();
                }
            }
        });
        shell.waitForIdle();
        shell.close();
    }

    static public void runCommand(List<String> commands) {
        final Interactive shell = openShell();
        shell.addCommand(commands, 0, new OnCommandResultListener() {
            @Override
            public void onCommandResult(int commandCode, int exitCode, List<String> output) {
                if (exitCode < 0) {
                    shell.close();
                }
            }
        });
        shell.waitForIdle();
        shell.close();
    }

    static public String runRootCommand(String command) {
        return runRootCommand(command, "u:r:init_shell:s0");
    }

    static public String runRootCommand(String command, String context) {
        if (!isRoot()) {
            return null;
        }
        final Interactive shell = openRootShell(context);
        StringBuilder sb = new StringBuilder();
        shell.addCommand(command, 0, new OnCommandResultListener() {
            @Override
            public void onCommandResult(int commandCode, int exitCode, List<String> output) {
                if (exitCode < 0) {
                    shell.close();
                } else {
                    //阿勒阿列
                }
            }
        });
        if (shell.waitForIdle()) {
            shell.close();
            return sb.toString();
        } else {
            shell.close();
            return null;

        }
    }
    static public String runRootCommand(String[] commands) {
        return runRootCommand(commands, "u:r:init_shell:s0");
    }

    static public String runRootCommand(String[] commands, String context) {
        if (!isRoot()) {
            return null;
        }
        final Interactive shell = openRootShell(context);
        StringBuilder sb = new StringBuilder();
        shell.addCommand(commands, 0, new OnCommandResultListener() {
            @Override
            public void onCommandResult(int commandCode, int exitCode, List<String> output) {
                if (exitCode < 0) {
                    shell.close();
                } else {
                    //阿勒阿列
                }
            }
        });
        if (shell.waitForIdle()) {
            shell.close();
            return sb.toString();
        } else {
            shell.close();
            return null;
        }
    }

    static public boolean isRoot() {
        return SU.available();
    }
    static public String mkCMD(String[] cmd){
        String c = "";
        for(int i = 0;i<cmd.length;i++)
           c+=cmd[i]+" ";
        return c;
    }
}


